File: python_pure/src/volarb_py/random_source.py
from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Final

PCG_MULTIPLIER: Final[int] = 0x2360ED051FC65DA44385DF649FCCF645
STATE_MASK: Final[int] = (1 << 128) - 1
WORD_MASK: Final[int] = (1 << 64) - 1
ROTATION_SHIFT: Final[int] = 122
MANTISSA_SHIFT: Final[int] = 11
MANTISSA_SCALE: Final[float] = 1.0 / 9007199254740992.0
TWO_PI: Final[float] = 6.283185307179586


class InvalidRandomSourceError(ValueError):
    pass


@dataclass(frozen=True)
class PcgState:
    state: int
    increment: int


@dataclass(frozen=True)
class NormalPair:
    state: PcgState
    first: float
    second: float


def advanced(source: PcgState) -> PcgState:
    return PcgState(
        state=(source.state * PCG_MULTIPLIER + source.increment) & STATE_MASK,
        increment=source.increment,
    )


def rotate_right(value: int, rotation: int) -> int:
    return ((value >> rotation) | (value << ((-rotation) & 63))) & WORD_MASK


def output_of(source: PcgState) -> int:
    xorshifted = ((source.state >> 64) ^ source.state) & WORD_MASK
    return rotate_right(xorshifted, source.state >> ROTATION_SHIFT)


def seeded_source(initial_state: int, sequence: int) -> PcgState:
    if initial_state < 0 or sequence < 0:
        raise InvalidRandomSourceError(f"a seed must not be negative, got {initial_state} and {sequence}")
    increment = ((sequence << 1) | 1) & STATE_MASK
    source = advanced(PcgState(state=0, increment=increment))
    source = PcgState(state=(source.state + initial_state) & STATE_MASK, increment=increment)
    return advanced(source)


def next_bits(source: PcgState) -> tuple[PcgState, int]:
    stepped = advanced(source)
    return stepped, output_of(stepped)


def next_uniform(source: PcgState) -> tuple[PcgState, float]:
    stepped, bits = next_bits(source)
    return stepped, (bits >> MANTISSA_SHIFT) * MANTISSA_SCALE


def next_positive_uniform(source: PcgState) -> tuple[PcgState, float]:
    stepped, bits = next_bits(source)
    return stepped, ((bits >> MANTISSA_SHIFT) + 1) * MANTISSA_SCALE


def next_standard_normal_pair(source: PcgState) -> NormalPair:
    after_first, first_uniform = next_positive_uniform(source)
    after_second, second_uniform = next_uniform(after_first)
    radius = math.sqrt(-2.0 * math.log(first_uniform))
    angle = TWO_PI * second_uniform
    return NormalPair(state=after_second, first=radius * math.cos(angle), second=radius * math.sin(angle))


def uniforms(source: PcgState, count: int) -> tuple[PcgState, list[float]]:
    if count < 0:
        raise InvalidRandomSourceError(f"count must not be negative, got {count}")
    values: list[float] = []
    current = source
    for _ in range(count):
        current, value = next_uniform(current)
        values.append(value)
    return current, values


def standard_normals(source: PcgState, count: int) -> tuple[PcgState, list[float]]:
    if count < 0:
        raise InvalidRandomSourceError(f"count must not be negative, got {count}")
    values: list[float] = []
    current = source
    while len(values) < count:
        pair = next_standard_normal_pair(current)
        current = pair.state
        values.append(pair.first)
        if len(values) < count:
            values.append(pair.second)
    return current, values

File: cpp_core/src/random_source.cpp
#include "volarb/random_source.hpp"

#include <cmath>

namespace volarb {

UnsignedWide pcg_multiplier() {
    return (static_cast<UnsignedWide>(0x2360ED051FC65DA4ULL) << 64) |
           static_cast<UnsignedWide>(0x4385DF649FCCF645ULL);
}

PcgState advanced(const PcgState& source) {
    return PcgState{source.state * pcg_multiplier() + source.increment, source.increment};
}

std::uint64_t rotate_right(std::uint64_t value, unsigned rotation) {
    return (value >> rotation) | (value << ((0U - rotation) & 63U));
}

std::uint64_t output_of(const PcgState& source) {
    const std::uint64_t xorshifted =
        static_cast<std::uint64_t>((source.state >> 64) ^ source.state);
    return rotate_right(xorshifted, static_cast<unsigned>(source.state >> rotation_shift));
}

PcgState seeded_source(UnsignedWide initial_state, UnsignedWide sequence) {
    const UnsignedWide increment = (sequence << 1) | 1;
    PcgState source = advanced(PcgState{0, increment});
    source = PcgState{source.state + initial_state, increment};
    return advanced(source);
}

BitsDraw next_bits(const PcgState& source) {
    const PcgState stepped = advanced(source);
    return BitsDraw{stepped, output_of(stepped)};
}

UniformDraw next_uniform(const PcgState& source) {
    const BitsDraw drawn = next_bits(source);
    return UniformDraw{drawn.state,
                       static_cast<double>(drawn.bits >> mantissa_shift) * mantissa_scale};
}

UniformDraw next_positive_uniform(const PcgState& source) {
    const BitsDraw drawn = next_bits(source);
    return UniformDraw{drawn.state,
                       static_cast<double>((drawn.bits >> mantissa_shift) + 1) * mantissa_scale};
}

NormalPair next_standard_normal_pair(const PcgState& source) {
    const UniformDraw first = next_positive_uniform(source);
    const UniformDraw second = next_uniform(first.state);
    const double radius = std::sqrt(-2.0 * std::log(first.value));
    const double angle = two_pi * second.value;
    return NormalPair{second.state, radius * std::cos(angle), radius * std::sin(angle)};
}

std::vector<double> uniforms(const PcgState& source, int count) {
    if (count < 0) {
        throw InvalidRandomSourceError("count must not be negative, got " + std::to_string(count));
    }
    std::vector<double> values;
    values.reserve(static_cast<std::size_t>(count));
    PcgState current = source;
    for (int index = 0; index < count; ++index) {
        const UniformDraw drawn = next_uniform(current);
        current = drawn.state;
        values.push_back(drawn.value);
    }
    return values;
}

std::vector<double> standard_normals(const PcgState& source, int count) {
    if (count < 0) {
        throw InvalidRandomSourceError("count must not be negative, got " + std::to_string(count));
    }
    std::vector<double> values;
    values.reserve(static_cast<std::size_t>(count));
    PcgState current = source;
    while (static_cast<int>(values.size()) < count) {
        const NormalPair pair = next_standard_normal_pair(current);
        current = pair.state;
        values.push_back(pair.first);
        if (static_cast<int>(values.size()) < count) {
            values.push_back(pair.second);
        }
    }
    return values;
}

}
